app/api/answers/generate_task_token.feature

```diff
@@ -0,0 +1,251 @@
+            "idAttempt": "102/0",
+            "idUser": "101",
+            "idItemLocal": "20",
+            "idItem": "task20",
+            "itemUrl": "http://taskplatform.mblockelet.info/task.html?taskId=403449543672183936",
+            "nbHintsGiven": "0",
+            "randomSeed": "6688673066740915448",
+            "platformName": "{{app().Config.GetString("token.platformName")}}",
+            "sLogin": "john"
+          }
+        },
+        "message": "created",
+        "success": true
+      }
+      """
+
+  Scenario: User is able to fetch a task token when participant is watched
+    Given I am the user with id "103"
+    When I send a POST request to "/answers/1/generate-task-token"
+    Then the response code should be 200
+    And the response body decoded as "GenerateTaskTokenResponse" should be, in JSON:
+      """
+      {
+        "data": {
+          "task_token": {
+            "date": "{{currentTimeInFormat("02-01-2006")}}",
+            "bAccessSolutions": false,
+            "bHintsAllowed": false,
+            "bIsAdmin": false,
+            "bReadAnswers": true,
+            "bSubmissionPossible": false,
+            "idAttempt": "101/0",
+            "idUser": "101",
+            "idItemLocal": "10",
+            "idItem": "task10",
+            "itemUrl": "http://taskplatform.mblockelet.info/task.html?taskId=403449543672183936",
+            "nbHintsGiven": "0",
+            "randomSeed": "2147886519731235493",
+            "platformName": "{{app().Config.GetString("token.platformName")}}",
+            "sLogin": "john"
+          }
+        },
+        "message": "created",
+        "success": true
+      }
+      """
+
+  Scenario: User is able to fetch a task token when participant is watched when having a started result on an other attempt
+    Given I am the user with id "103"
+    When I send a POST request to "/answers/3/generate-task-token"
+    Then the response code should be 200
+    And the response body decoded as "GenerateTaskTokenResponse" should be, in JSON:
+      """
+      {
+        "data": {
+          "task_token": {
+            "date": "{{currentTimeInFormat("02-01-2006")}}",
+            "bAccessSolutions": false,
+            "bHintsAllowed": false,
+            "bIsAdmin": false,
+            "bReadAnswers": true,
+            "bSubmissionPossible": false,
+            "idAttempt": "101/0",
+            "idUser": "101",
+            "idItemLocal": "30",
+            "idItem": "task30",
+            "itemUrl": "http://taskplatform.mblockelet.info/task.html?taskId=403449543672183936",
+            "nbHintsGiven": "0",
+            "randomSeed": "2147886519731235493",
+            "platformName": "{{app().Config.GetString("token.platformName")}}",
+            "sLogin": "john"
+          }
+        },
+        "message": "created",
+        "success": true
+      }
+      """
+
+  Scenario: bAccessSolutions is true when item `can_view` >= 'solution'
```

Comment:
For `bAccessSolutions` (both tests): as the user of the answer is the same as the user making the request, I am not able to spot an error is perm checking (typically checking the perm of the wrong user)

It is strange to write "when item `can_view` >= 'solution'" ... it is the user who has a permission... and here it should even be written that it is the requester.
Same remark for the next test "when item has been validated" -> we have to know that it is by the requester... (it should not be the result related to the answer)

Reply:
fixed, the permissions checks in the code where wrong indeed :)